File: app/src/main/java/com/persona5dex/viewmodels/PersonaMainListViewModel.java
package com.persona5dex.viewmodels;

import androidx.annotation.NonNull;
import androidx.arch.core.util.Function;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.persona5dex.ArcanaNameProvider;
import com.persona5dex.PersonaFilters;
import com.persona5dex.models.Enumerations;
import com.persona5dex.models.GameType;
import com.persona5dex.models.MainListPersona;
import com.persona5dex.models.PersonaFilterArgs;
import com.persona5dex.models.room.PersonaShadowName;
import com.persona5dex.repositories.MainPersonaRepository;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * Created by Rechee on 11/18/2017.
 */

public class PersonaMainListViewModel extends ViewModel{
    private final ArcanaNameProvider arcanaNameProvider;

    private MutableLiveData<String> personaSearchName;
    private MutableLiveData<Boolean> personasByName;
    private MutableLiveData<Boolean> personasByLevel;
    private MutableLiveData<Boolean> personasByArcana;
    private MutableLiveData<PersonaFilterArgs> personaFilterArgs;

    private MainPersonaRepository repository;

    private Comparator<MainListPersona> sortByPersonaNameDesc;
    private Comparator<MainListPersona> sortByPersonaNameAsc;
    private Comparator<MainListPersona> sortByPersonaLevelAsc;
    private Comparator<MainListPersona> sortByPersonaLevelDesc;
    private Comparator<MainListPersona> sortByPersonaArcanaAsc;
    private Comparator<MainListPersona> sortByPersonaArcanaDesc;
    private LiveData<List<MainListPersona>> allPersonas;

    private LiveData<List<MainListPersona>> filteredPersonas;

    public PersonaMainListViewModel(final MainPersonaRepository repository, ArcanaNameProvider arcanaNameProvider){
        this.repository = repository;
        this.arcanaNameProvider = arcanaNameProvider;

        allPersonas = repository.getAllPersonasForMainList();
        allPersonas.observeForever(mainListPersonas -> {

        });

        filteredPersonas = Transformations.switchMap(allPersonas, personas -> {
            MediatorLiveData<List<MainListPersona>> filteredLiveData = new MediatorLiveData<>();

            filteredLiveData.addSource(personaSearchName, searchValue -> {
                if(searchValue == null || searchValue.isEmpty()){
                    filteredLiveData.setValue(personas);
                }
                else{
                    List<MainListPersona> finalList = new ArrayList<>();

                    for (MainListPersona mainListPersona : personas) {
                        final String searchValueLower = searchValue.toLowerCase();

                        if(mainListPersona.name.toLowerCase().contains(searchValueLower)){
                            finalList.add(mainListPersona);
                        }
                        else{
                            for (PersonaShadowName personaShadowName : mainListPersona.personaShadowNames) {
                                if(personaShadowName.shadowName.toLowerCase().contains(searchValueLower)){
                                    finalList.add(mainListPersona);
                                    break;
                                }
                            }
                        }
                    }

                    filteredLiveData.setValue(finalList);
                }
            });

            filteredLiveData.addSource(personasByName, asc -> {
                if(asc == null){
                    asc = true;
                }

                List<MainListPersona> newPersonas = filteredPersonas.getValue();
                if(newPersonas != null){
                    if(newPersonas.size() == 1){
                        return;
                    }

                    if(asc){
                        Collections.sort(newPersonas, sortByPersonaNameAsc);
                    }
                    else{
                        Collections.sort(newPersonas, sortByPersonaNameDesc);
                    }
                }

                filteredLiveData.setValue(newPersonas);
            });

            filteredLiveData.addSource(personasByLevel, asc -> {
                if(asc == null){
                    asc = true;
                }

                final List<MainListPersona> newPersonas = filteredPersonas.getValue();

                if(newPersonas != null){
                    if(newPersonas.size() == 1){
                        return;
                    }

                    if(asc){
                        Collections.sort(newPersonas, sortByPersonaLevelAsc);
                    }
                    else{
                        Collections.sort(newPersonas, sortByPersonaLevelDesc);
                    }
                }

                filteredLiveData.setValue(newPersonas);
            });

            filteredLiveData.addSource(personasByArcana, asc -> {
                if(asc == null){
                    asc = true;
                }

                List<MainListPersona> newPersonas = filteredLiveData.getValue();

                if(newPersonas != null){
                    if(newPersonas.size() == 1){
                        return;
                    }

                    if(asc){
                        Collections.sort(newPersonas, sortByPersonaArcanaAsc);
                    }
                    else{
                        Collections.sort(newPersonas, sortByPersonaArcanaDesc);
                    }
                }

                filteredLiveData.setValue(newPersonas);
            });

            filteredLiveData.addSource(personaFilterArgs, (final PersonaFilterArgs inputFilterArgs) -> {

                PersonaFilterArgs filterArgs = inputFilterArgs;

                if(filterArgs == null){
                    filterArgs = new PersonaFilterArgs();
                }

                List<MainListPersona> finalValue = new ArrayList<>();

                if(personas != null) {
                    final List<MainListPersona> newPersonas = PersonaFilters.filterGameType(personas, inputFilterArgs.gameType);

                    for (MainListPersona persona : newPersonas) {

                        if (persona.rare && !filterArgs.rarePersona) {
                            continue;
                        }

                        if (persona.dlc && !filterArgs.dlcPersona) {
                            continue;
                        }

                        if (filterArgs.arcana != Enumerations.Arcana.ANY && persona.arcana != filterArgs.arcana) {
                            continue;
                        }

                        if (persona.level < filterArgs.minLevel || persona.level > filterArgs.maxLevel) {
                            continue;
                        }

                        finalValue.add(persona);
                    }
                }

                filteredLiveData.setValue(finalValue);
            });

            return filteredLiveData;
        });

        personaSearchName = new MutableLiveData<>();
        personasByName = new MutableLiveData<>();
        personasByLevel = new MutableLiveData<>();
        personasByArcana = new MutableLiveData<>();
        personaFilterArgs = new MutableLiveData<>();

        sortByPersonaNameAsc = (o1, o2) -> o1.name.compareTo(o2.name);

        sortByPersonaNameDesc = (o1, o2) -> o1.name.compareTo(o2.name) * -1;

        sortByPersonaLevelAsc = (o1, o2) -> Integer.compare(o1.level, o2.level);
        sortByPersonaLevelDesc = (o1, o2) -> Integer.compare(o1.level, o2.level) * -1;
        sortByPersonaArcanaAsc = (o1, o2) -> arcanaNameProvider.getArcanaNameForDisplay(o1.arcana).compareTo(arcanaNameProvider.getArcanaNameForDisplay(o2.arcana));
        sortByPersonaArcanaDesc = (o1, o2) -> arcanaNameProvider.getArcanaNameForDisplay(o1.arcana).compareTo(arcanaNameProvider.getArcanaNameForDisplay(o2.arcana)) * -1;
    }

    public void filterPersonas(final String personaNameQuery){
        this.personaSearchName.postValue(personaNameQuery);
    }

    public LiveData<List<MainListPersona>> getFilteredPersonas() {
        return filteredPersonas;
    }

    public void sortPersonasByName(final boolean asc){
        personasByName.setValue(asc);
    }

    public void sortPersonasByLevel(final boolean asc){
        personasByLevel.setValue(asc);
    }

    public void sortPersonasByArcana(boolean asc) {
        personasByArcana.setValue(asc);
    }

    public void filterPersonas(final PersonaFilterArgs filterArgs) {
        personaFilterArgs.setValue(filterArgs);
    }

    public void filterPersonas(@NonNull GameType gameType) {
        PersonaFilterArgs newFilterArgs  = personaFilterArgs.getValue();
        if(newFilterArgs == null){
            newFilterArgs = new PersonaFilterArgs();
        }
        newFilterArgs.gameType = gameType;
        personaFilterArgs.setValue(newFilterArgs);
    }
}

File: app/src/main/java/com/persona5dex/viewmodels/PersonaMainListViewModel.kt
package com.persona5dex.viewmodels

import android.content.SharedPreferences
import androidx.lifecycle.*
import com.persona5dex.ArcanaNameProvider
import com.persona5dex.SHARED_PREF_KEY_GAME_TYPE
import com.persona5dex.extensions.toLowerCaseInsensitive
import com.persona5dex.filterGameType
import com.persona5dex.models.*
import com.persona5dex.models.GameType.Companion.toGameType
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import java.util.*

/**
 * Created by Rechee on 11/18/2017.
 */
class PersonaMainListViewModel(
        private val arcanaNameProvider: ArcanaNameProvider,
        private var currentGameType: GameType,
        private val personaRepository: PersonaRepository,
        private val defaultSharedSharedPreferences: SharedPreferences
) : ViewModel(), SharedPreferences.OnSharedPreferenceChangeListener {
    private val personaSearchName: MutableLiveData<String> = MutableLiveData()
    private val personasByName: MutableLiveData<Boolean> = MutableLiveData()
    private val personasByLevel: MutableLiveData<Boolean> = MutableLiveData()
    private val personasByArcana: MutableLiveData<Boolean> = MutableLiveData()
    private val personaFilterArgs: MutableLiveData<PersonaFilterArgs> = MutableLiveData()
    private val stateLiveData = MutableLiveData<State>()
    private val gameTypeLiveData = MutableLiveData<GameType>()
    private val allPersonas by lazy {
        viewModelScope.async {
            personaRepository.getPersonas()
        }
    }

    private val filteredLiveData = MediatorLiveData<List<MainListPersona>>().apply {
        addSource(personaSearchName) { searchValue: String? ->
            viewModelScope.launch {
                try {
                    val allPersonas = getNewPersonaList(checkNotNull(personaFilterArgs.value))
                    if (searchValue.isNullOrEmpty()) {
                        setValue(allPersonas)
                    } else {
                        val finalList: MutableList<MainListPersona> = ArrayList()
                        for (mainListPersona in allPersonas) {
                            val searchValueLower = searchValue.toLowerCaseInsensitive()
                            if (mainListPersona.name.toLowerCaseInsensitive().contains(searchValueLower)) {
                                finalList.add(mainListPersona)
                            } else {
                                for (personaShadowName in mainListPersona.personaShadowNames) {
                                    if (personaShadowName.shadowName.toLowerCaseInsensitive().contains(searchValueLower)) {
                                        finalList.add(mainListPersona)
                                        break
                                    }
                                }
                            }
                        }
                        setValue(finalList)
                    }
                } finally {
                    finishLoading()
                }
            }
        }
        addSource(personasByName) { asc: Boolean? ->
            val newPersonas = value
            if (newPersonas != null) {
                if (newPersonas.size == 1) {
                    return@addSource
                }
                lastPersonaComparator = if (asc != false) {
                    Collections.sort(newPersonas, sortByPersonaNameAsc)
                    sortByPersonaNameAsc
                } else {
                    Collections.sort(newPersonas, sortByPersonaNameDesc)
                    sortByPersonaNameDesc
                }
            }
            value = newPersonas
            finishLoading()
        }
        addSource(personasByLevel) { asc: Boolean? ->
            val newPersonas = value
            if (newPersonas != null) {
                if (newPersonas.size == 1) {
                    return@addSource
                }
                lastPersonaComparator = if (asc != false) {
                    Collections.sort(newPersonas, sortByPersonaLevelAsc)
                    sortByPersonaLevelAsc
                } else {
                    Collections.sort(newPersonas, sortByPersonaLevelDesc)
                    sortByPersonaLevelDesc
                }
            }
            value = newPersonas
            finishLoading()
        }
        addSource(personasByArcana) { asc: Boolean? ->
            val newPersonas = value
            if (newPersonas != null) {
                if (newPersonas.size == 1) {
                    return@addSource
                }
                lastPersonaComparator = if (asc != false) {
                    Collections.sort(newPersonas, sortByPersonaArcanaAsc)
                    sortByPersonaArcanaAsc
                } else {
                    Collections.sort(newPersonas, sortByPersonaArcanaDesc)
                    sortByPersonaArcanaDesc
                }
            }
            value = newPersonas
            finishLoading()
        }
        addSource(personaFilterArgs) { inputFilterArgs: PersonaFilterArgs ->
            val finalValue: MutableList<MainListPersona> = ArrayList()
            viewModelScope.launch {
                try {
                    val newPersonas = getNewPersonaList(inputFilterArgs)

                    for (persona in newPersonas) {
                        if (persona.dlc && !inputFilterArgs.dlcPersona) {
                            continue
                        }
                        if (inputFilterArgs.arcana != Enumerations.Arcana.ANY && persona.arcana != inputFilterArgs.arcana) {
                            continue
                        }
                        if (persona.level < inputFilterArgs.minLevel || persona.level > inputFilterArgs.maxLevel) {
                            continue
                        }
                        finalValue.add(persona)
                    }
                    setValue(finalValue)
                } finally {
                    finishLoading()
                }
            }
        }
    }

    private suspend fun getNewPersonaList(inputFilterArgs: PersonaFilterArgs): List<MainListPersona> {
        val comparator = lastPersonaComparator ?: sortByPersonaNameAsc
        val allPersonas = if (allPersonas.isCompleted) allPersonas.getCompleted() else allPersonas.await()
        return allPersonas
                .filterGameType(currentGameType, inputFilterArgs.basePersonas, inputFilterArgs.royalPersonas)
                .sortedWith(comparator)
    }

    private val sortByPersonaNameDesc: Comparator<MainListPersona>
    private val sortByPersonaNameAsc: Comparator<MainListPersona>
    private val sortByPersonaLevelAsc: Comparator<MainListPersona>
    private val sortByPersonaLevelDesc: Comparator<MainListPersona>
    private val sortByPersonaArcanaAsc: Comparator<MainListPersona>
    private val sortByPersonaArcanaDesc: Comparator<MainListPersona>

    private var lastPersonaComparator: Comparator<MainListPersona>? = null

    init {
        sortByPersonaNameAsc = Comparator { o1: MainListPersona, o2: MainListPersona -> o1.name.compareTo(o2.name) }
        sortByPersonaNameDesc = Comparator { o1: MainListPersona, o2: MainListPersona -> o1.name.compareTo(o2.name) * -1 }
        sortByPersonaLevelAsc = Comparator { o1: MainListPersona, o2: MainListPersona -> o1.level.compareTo(o2.level) }
        sortByPersonaLevelDesc = Comparator { o1: MainListPersona, o2: MainListPersona -> o1.level.compareTo(o2.level) * -1 }
        sortByPersonaArcanaAsc = Comparator { o1: MainListPersona, o2: MainListPersona -> getPersonaNameComparison(o1, o2) }
        sortByPersonaArcanaDesc = Comparator { o1: MainListPersona, o2: MainListPersona -> getPersonaNameComparison(o1, o2) * -1 }

        personaFilterArgs.value = PersonaFilterArgs().apply {
            basePersonas = true
            royalPersonas = currentGameType == GameType.ROYAL
        }

        filterPersonas(currentGameType)
        stateLiveData.postValue(State.LoadingStarted)

        defaultSharedSharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    fun getState(): LiveData<State> = stateLiveData

    fun filterPersonas(personaNameQuery: String?) {
        startLoading()
        personaSearchName.postValue(personaNameQuery)
    }

    fun sortPersonasByName(asc: Boolean) {
        startLoading()
        personasByName.value = asc
    }

    fun sortPersonasByLevel(asc: Boolean) {
        startLoading()
        personasByLevel.value = asc
    }

    fun sortPersonasByArcana(asc: Boolean) {
        startLoading()
        personasByArcana.value = asc
    }

    fun filterPersonas(filterArgs: PersonaFilterArgs) {
        startLoading()
        personaFilterArgs.value = filterArgs
    }

    fun filterPersonas(gameType: GameType) {
        startLoading()
        var gamTypeChanged = false
        if (gameType != currentGameType) {
            gamTypeChanged = true
            gameTypeLiveData.value = gameType
        }
        currentGameType = gameType

        val newFilterArgs = checkNotNull(personaFilterArgs.value) {
            "not expecting filter args to be null. game type $gameType"
        }.let {
            if (gamTypeChanged) {
                if (gameType == GameType.BASE) {
                    it.basePersonas = true
                    it.royalPersonas = false
                } else {
                    it.basePersonas = true
                    it.royalPersonas = true
                }
            }
            it
        }

        filterPersonas(newFilterArgs)
    }

    fun getGameType() = gameTypeLiveData

    fun getFilteredPersonas() = filteredLiveData

    fun getFilterArgs() = personaFilterArgs

    fun getArcanaNamesForSpinner() = arcanaNameProvider.getAllArcanaNamesForDisplay()

    private fun startLoading() {
        stateLiveData.value = State.LoadingStarted
    }

    private fun finishLoading() {
        stateLiveData.value = State.LoadingCompleted
    }

    override fun onCleared() {
        super.onCleared()
        defaultSharedSharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    sealed class State {
        object LoadingStarted : State()
        object LoadingCompleted : State()
    }

    private fun getPersonaNameComparison(personaOne: MainListPersona, personaTwo: MainListPersona) =
            arcanaNameProvider.getArcanaNameForDisplay(personaOne.arcana).compareTo(arcanaNameProvider.getArcanaNameForDisplay(personaTwo.arcana))

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        if (key == SHARED_PREF_KEY_GAME_TYPE) {
            val newGameType = defaultSharedSharedPreferences.getInt(SHARED_PREF_KEY_GAME_TYPE, GameType.BASE.value).toGameType()
            if (newGameType != gameTypeLiveData.value) {
                gameTypeLiveData.value = newGameType
            }
        }
    }
}